# Part 1. Cleaning Patient Lab Ranges - David Dangond


In [1]:
from google.colab import drive
import pandas as pd
import numpy as np
drive.mount('/content/drive')
filepath = "/content/drive/MyDrive/Spoke/"

Mounted at /content/drive


In [ ]:
not_clean = pd.read_csv(filepath + 'OCT20_range_w_demo_and_vals.tsv', sep='\t')

In [ ]:
# FOR MS RESULTS
res = pd.read_table(filepath + 'matched_lab_df.tsv')


<ipython-input-17-74593fd8aaa6>:2: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  res = pd.read_table(filepath + 'matched_lab_df.tsv')


In [ ]:
not_clean

measurement_concept_id  value_as_number  value_as_concept_id  \
0                        3010421             7.34                  NaN   
1                        3013502            69.00                  NaN   
2                        3013502            72.00                  NaN   
3                        3027315            28.00                  NaN   
4                        3027315            31.00                  NaN   
...                          ...              ...                  ...   
14244781                 3016723             1.30                  NaN   
14244782                 3000905            12.60                  NaN   
14244783                 3046553             0.41                  NaN   
14244784                 3004327             3.35                  NaN   
14244785                 3023996             0.00                  NaN   

          unit_concept_id measurement_source_value unit_source_value  \
0                     NaN                  LAB5808               NaN   
1                  8554.0                  LAB5808                 %   
2                  8554.0                  LAB5809                 %   
3                     NaN                  LAB5808             mm Hg   
4                     NaN                  LAB5809             mm Hg   
...                   ...                      ...               ...   
14244781           8840.0                  LAB3001             mg/dL   
14244782              NaN                   LAB294           x10E9/L   
14244783              NaN                   LAB293           x10E9/L   
14244784              NaN                   LAB293           x10E9/L   
14244785              NaN                   LC3743    mmol/mol creat   

          range_low  range_high gender_source_value  ethnicity_source_value  \
0              7.35        7.45              Female  Not Hispanic or Latino   
1             95.00       99.00              Female  Not Hispanic or Latino   
2             95.00       99.00              Female  Not Hispanic or Latino   
3             80.00      100.00              Female  Not Hispanic or Latino   
4             80.00      100.00              Female  Not Hispanic or Latino   
...             ...         ...                 ...                     ...   
14244781       0.61        1.24                Male      Hispanic or Latino   
14244782       3.40       10.00              Female                 Unknown   
14244783       0.00        0.20              Female      Hispanic or Latino   
14244784       0.90        2.90                Male  Not Hispanic or Latino   
14244785       1.00       14.00                Male      Hispanic or Latino   

           race_source_value  age_at_test  
0                      Asian            0  
1                      Asian            0  
2                      Asian            0  
3                      Asian            0  
4                      Asian            0  
...                      ...          ...  
14244781  White or Caucasian           53  
14244782             Unknown           71  
14244783  White or Caucasian            2  
14244784  White or Caucasian           36  
14244785             Unknown           13  

[14244786 rows x 12 columns]

In [ ]:
len(set(res[~res['OMOPid'].isin(list(set(not_clean['measurement_concept_id'])))]['OMOPid']))

1276

In [ ]:
len(set(res['OMOPid']))

1941

In [ ]:
len(set(not_clean['measurement_concept_id']))

983

In [ ]:
# ms = not_clean[not_clean['measurement_concept_id'].isin(list(set(res['OMOPid'])))]
ms = res[res['OMOPid'].isin(list(set(not_clean['measurement_concept_id'])))]

print('number of labs from original data set included in MS Dataset: ', len(set(ms['OMOPid'])))


not_ms = not_clean[~not_clean['measurement_concept_id'].isin(list(set(res['OMOPid'])))]

print('number of labs from original data set not included in MS Dataset: ', len(set(not_ms['measurement_concept_id'])))

number of labs from original data set included in MS Dataset:  665
number of labs from original data set not included in MS Dataset:  318


In [ ]:
not_clean

measurement_concept_id  value_as_number  value_as_concept_id  \
0                        3010421             7.34                  NaN   
1                        3013502            69.00                  NaN   
2                        3013502            72.00                  NaN   
3                        3027315            28.00                  NaN   
4                        3027315            31.00                  NaN   
...                          ...              ...                  ...   
14244781                 3016723             1.30                  NaN   
14244782                 3000905            12.60                  NaN   
14244783                 3046553             0.41                  NaN   
14244784                 3004327             3.35                  NaN   
14244785                 3023996             0.00                  NaN   

          unit_concept_id measurement_source_value unit_source_value  \
0                     NaN                  LAB5808               NaN   
1                  8554.0                  LAB5808                 %   
2                  8554.0                  LAB5809                 %   
3                     NaN                  LAB5808             mm Hg   
4                     NaN                  LAB5809             mm Hg   
...                   ...                      ...               ...   
14244781           8840.0                  LAB3001             mg/dL   
14244782              NaN                   LAB294           x10E9/L   
14244783              NaN                   LAB293           x10E9/L   
14244784              NaN                   LAB293           x10E9/L   
14244785              NaN                   LC3743    mmol/mol creat   

          range_low  range_high gender_source_value  ethnicity_source_value  \
0              7.35        7.45              Female  Not Hispanic or Latino   
1             95.00       99.00              Female  Not Hispanic or Latino   
2             95.00       99.00              Female  Not Hispanic or Latino   
3             80.00      100.00              Female  Not Hispanic or Latino   
4             80.00      100.00              Female  Not Hispanic or Latino   
...             ...         ...                 ...                     ...   
14244781       0.61        1.24                Male      Hispanic or Latino   
14244782       3.40       10.00              Female                 Unknown   
14244783       0.00        0.20              Female      Hispanic or Latino   
14244784       0.90        2.90                Male  Not Hispanic or Latino   
14244785       1.00       14.00                Male      Hispanic or Latino   

           race_source_value  age_at_test  
0                      Asian            0  
1                      Asian            0  
2                      Asian            0  
3                      Asian            0  
4                      Asian            0  
...                      ...          ...  
14244781  White or Caucasian           53  
14244782             Unknown           71  
14244783  White or Caucasian            2  
14244784  White or Caucasian           36  
14244785             Unknown           13  

[14244786 rows x 12 columns]

In [ ]:
ms

cohort_id   OMOPid   omop_date  value_as_number  value_as_concept_id  \
0                0  3013721  2009-09-03            18.00                  NaN   
1                0  3019550  2009-02-10           138.00                  NaN   
2                0  3035995  2009-09-03            59.00                  NaN   
3                0  3044491  2009-02-10           122.00                  NaN   
4                0  3022192  2009-02-10            50.00                  NaN   
...            ...      ...         ...              ...                  ...   
1294817      41177  3000905  1998-02-26             2.10                  NaN   
1294818      41177  3015884  1998-08-14           227.00                  NaN   
1294819      41177  3004501  1998-08-14            88.00                  NaN   
1294820      41177  3023314  1998-02-26            43.70                  NaN   
1295022      41241  3009201  2018-05-17             0.54                  NaN   

         unit_concept_id unit_source_value  measurement_source_value  \
0                 8645.0               U/L                   1526000   
1                 8753.0            MMOL/L                   1534098   
2                 8645.0               U/L                   1525848   
3                 8840.0             MG/DL                       676   
4                 8840.0             MG/DL                   1552156   
...                  ...               ...                       ...   
1294817           9444.0           x10E9/L                      1004   
1294818           8837.0             ug/dL                   1900562   
1294819           8840.0             mg/dL                   1510655   
1294820           8554.0           PERCENT                   1534436   
1295022           9040.0             mIU/L                      2465   

        value_source_value  
0                       18  
1                      138  
2                       59  
3                      122  
4                       50  
...                    ...  
1294817                2.1  
1294818                227  
1294819                 88  
1294820               43.7  
1295022               0.54  

[660445 rows x 9 columns]

In [ ]:
renamed = not_clean.rename(columns = {'measurement_concept_id': 'OMOPid', 'measurement_source_value' : 'manufacturer', 'unit_source_value':'unit', 'gender_source_value':'gender','ethnicity_source_value':'ethnicity','race_source_value':'race','age_at_test':'age'})
filtered = renamed[['OMOPid', 'manufacturer', 'unit', 'range_low', 'range_high', 'gender', 'ethnicity', 'race', 'age']]







In [ ]:
omc_raw = pd.read_csv(filepath + 'old_measurement_concepts.tsv', sep='\t')
omc = omc_raw.rename(columns={'concept_id': "OMOPid", 'concept_code': 'LOINCid'})

In [ ]:
print('old_measurement_concepts:')
# not useful, many labs done multiple times
# omcUnique = omc.drop_duplicates(subset = ['OMOPid','LOINCid'])
# print(len(omcUnique))
# print(len(omc))
display(omc.head())

old_measurement_concepts:


OMOPid                                       concept_name    domain_id  \
0  44786677         Genetic risk algorithm sensitivity [Ratio]  Measurement   
1  44786872  Fondaparinux [Mass/volume] in Platelet poor pl...  Measurement   
2  44786663  Serotonin release 0.5 IU/mL heparin.unfraction...  Measurement   
3  44786754      Tacrolimus [Mass/volume] in Blood by LC/MS/MS  Measurement   
4  42528717  Cashew nut recombinant (rAna o) 3 IgE Ab [Unit...  Measurement   

  vocabulary_id      concept_class_id standard_concept  LOINCid  \
0         LOINC  Clinical Observation                S  74020-9   
1         LOINC              Lab Test                S  74217-1   
2         LOINC              Lab Test                S  74005-0   
3         LOINC              Lab Test                S  74097-7   
4         LOINC              Lab Test                S  82539-8   

  valid_start_date valid_end_date invalid_reason  
0       2013-07-23     2099-12-31            NaN  
1       2013-10-28     2099-12-31            NaN  
2       2013-07-22     2099-12-31            NaN  
3       2013-08-29     2099-12-31            NaN  
4       2017-02-21     2099-12-31            NaN

In [ ]:
# create map to later add LOINCids to the 2020 demographics DF
loincMap = dict(zip(list(omc['OMOPid']), list(omc['LOINCid'])))
loincMap

{44786677: '74020-9',
 44786872: '74217-1',
 44786663: '74005-0',
 44786754: '74097-7',
 42528717: '82539-8',
 46235951: '77741-7',
 46236302: '77636-9',
 43055684: '72824-6',
 43055120: '72493-0',
 43055151: '72524-2',
 43055153: '72526-7',
 43055279: '72314-8',
 43055441: '72607-5',
 43055122: '72495-5',
 43055144: '72517-6',
 43055457: '72623-2',
 43055031: '72400-5',
 43055560: '72726-3',
 43055562: '72728-9',
 43055033: '72402-1',
 43055022: '72391-6',
 43055121: '72494-8',
 44786773: '74116-5',
 21493555: '80674-5',
 21493565: '80684-4',
 40761514: '58413-6',
 40761531: '58430-0',
 40760859: '57749-4',
 40760911: '57802-1',
 40761010: '57901-1',
 46237016: '77174-1',
 46235436: '75654-4',
 46236749: '77039-6',
 46236971: '78002-3',
 40760439: '57320-4',
 46235690: '75882-1',
 46235955: '77745-8',
 46235190: '76651-9',
 46236293: '77627-8',
 40762156: '59063-8',
 46237030: '77207-9',
 40762529: '59466-3',
 40765679: '62947-7',
 40765841: '63109-3',
 40765889: '63157-2',
 40766283:

In [ ]:
# add Loinc IDs to demographic DF
filtered['LOINCid'] = [loincMap[filtered['OMOPid'][omop]] for omop in range(len(filtered))]
filtered

<ipython-input-29-4dd0c18e106e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered['LOINCid'] = [loincMap[filtered['OMOPid'][omop]] for omop in range(len(filtered))]


OMOPid manufacturer            unit  range_low  range_high  gender  \
0         3010421      LAB5808             NaN       7.35        7.45  Female   
1         3013502      LAB5808               %      95.00       99.00  Female   
2         3013502      LAB5809               %      95.00       99.00  Female   
3         3027315      LAB5808           mm Hg      80.00      100.00  Female   
4         3027315      LAB5809           mm Hg      80.00      100.00  Female   
...           ...          ...             ...        ...         ...     ...   
14244781  3016723      LAB3001           mg/dL       0.61        1.24    Male   
14244782  3000905       LAB294         x10E9/L       3.40       10.00  Female   
14244783  3046553       LAB293         x10E9/L       0.00        0.20  Female   
14244784  3004327       LAB293         x10E9/L       0.90        2.90    Male   
14244785  3023996       LC3743  mmol/mol creat       1.00       14.00    Male   

                       ethnicity                race  age  LOINCid  
0         Not Hispanic or Latino               Asian    0  11558-4  
1         Not Hispanic or Latino               Asian    0  20564-1  
2         Not Hispanic or Latino               Asian    0  20564-1  
3         Not Hispanic or Latino               Asian    0  11556-8  
4         Not Hispanic or Latino               Asian    0  11556-8  
...                          ...                 ...  ...      ...  
14244781      Hispanic or Latino  White or Caucasian   53   2160-0  
14244782                 Unknown             Unknown   71   6690-2  
14244783      Hispanic or Latino  White or Caucasian    2  43743-4  
14244784  Not Hispanic or Latino  White or Caucasian   36    731-0  
14244785      Hispanic or Latino             Unknown   13  25132-2  

[14244786 rows x 10 columns]

In [ ]:
print("range is 0-0 for the following units:")
set(filtered[(filtered['range_low']== 0) & (filtered['range_high']== 0)]['unit'])

range is 0-0 for the following units:


{'%', 'PERCENT', 'mmol/mol creat'}

In [ ]:
# convert units to lowercase
filtered['unit'] = [str(unit).lower().strip() for unit in filtered['unit']]
filtered

<ipython-input-31-787c85b7c6d8>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered['unit'] = [str(unit).lower().strip() for unit in filtered['unit']]


OMOPid manufacturer            unit  range_low  range_high  gender  \
0         3010421      LAB5808             nan       7.35        7.45  Female   
1         3013502      LAB5808               %      95.00       99.00  Female   
2         3013502      LAB5809               %      95.00       99.00  Female   
3         3027315      LAB5808           mm hg      80.00      100.00  Female   
4         3027315      LAB5809           mm hg      80.00      100.00  Female   
...           ...          ...             ...        ...         ...     ...   
14244781  3016723      LAB3001           mg/dl       0.61        1.24    Male   
14244782  3000905       LAB294         x10e9/l       3.40       10.00  Female   
14244783  3046553       LAB293         x10e9/l       0.00        0.20  Female   
14244784  3004327       LAB293         x10e9/l       0.90        2.90    Male   
14244785  3023996       LC3743  mmol/mol creat       1.00       14.00    Male   

                       ethnicity                race  age  LOINCid  
0         Not Hispanic or Latino               Asian    0  11558-4  
1         Not Hispanic or Latino               Asian    0  20564-1  
2         Not Hispanic or Latino               Asian    0  20564-1  
3         Not Hispanic or Latino               Asian    0  11556-8  
4         Not Hispanic or Latino               Asian    0  11556-8  
...                          ...                 ...  ...      ...  
14244781      Hispanic or Latino  White or Caucasian   53   2160-0  
14244782                 Unknown             Unknown   71   6690-2  
14244783      Hispanic or Latino  White or Caucasian    2  43743-4  
14244784  Not Hispanic or Latino  White or Caucasian   36    731-0  
14244785      Hispanic or Latino             Unknown   13  25132-2  

[14244786 rows x 10 columns]

# Situations in which labs tests don't have clear ranges
1. range is not a number

    - DONE Result: All are floats.


2. range does not have reasonable low or high value

    2a) for [low, high], low > high
    
        - DONE Result: 121 rows (err_range)
        
    2b) low < 0 or high < 0
    
        - DONE Result: check passed (0 match this) (neg_low_high)
    
3. range is different for same manufacturer, omopid, and demographic

        - DONE Progress: There exist ranges that are different despite being from same manufacturer, omopid, and demographic.
            - Charlotte's suggestion: Could add year it was taken in case the range was updated later on. However, might be a better option to just add a confidence interval.
            - my approach: Apply confidence interval. If difference is beyond the CI, add both to a separate unclear ranges table ("conflicting_ranges_df", so its easier to find what about it makes it unclear.
    ***Progress:***

                - implemented function to convert to "any" if meets condition (condition being that all values for this demographic are present and the range is the same)
                    - need to check with Charlotte if this is the way she intends this to be implemented (DONE)
  
                   
        
4. unit is unclear (nan,  'not reported', 'index', 'units', 'undex', 'index value', '(calc)')
    - DONE Result: 372,650 rows
    
5. age is negative
    - DONE Result: 90 rows

In [ ]:
sum(list(np.isnan(filtered['range_high'].values)))

0

## Situations in which lab tests don't have clear ranges (CODE)

1. range is not a number

    - DONE Result: All are floats.

In [ ]:
# 1. range is not a float & not NaN
typelow = sum(list(filtered['range_low'].map(type) != float) or list(np.isnan(filtered['range_low'].values)))
typehigh = sum(list(filtered['range_high'].map(type) != float) or list(np.isnan(filtered['range_high'].values)))
print('Number of rows where range_low is not a float or NaN:\n', typelow)
print('Number of rows where range_high is not a float or NaN:\n', typehigh)
# all are floats.

Number of rows where range_low is not a float or NaN:
 0
Number of rows where range_high is not a float or NaN:
 0


2. range does not have reasonable low or high value

    2a) for [low, high], low > high
    
        - DONE Result: 121 rows (errRange)
        
    2b) low < 0 or high < 0
    
        - DONE Result: check passed (0 match this) (neg_low_high)

In [ ]:
# 2a. for [low, high], low >= high
err_range = filtered[filtered['range_low'] > filtered['range_high']]
# append unclear indeces to to_drop
to_drop = list(err_range.index)

print("Number of rows where range_low > range_high:\n", str(len(err_range)))


# 2b. low or high < 0
neg_low_high = filtered[(filtered['range_low'] < 0) | (filtered['range_high'] < 0)]

print("Number of rows where range_low < 0 or range_high < 0:\n", str(len(neg_low_high)))


# Dont need to check if range_high = 0, since a low, high range = 0,0 is valid in cases where patients should have 0 concentration of something.
# invalid_high = filtered[(filtered['range_high'] <= 0)]
# print("Number of rows where range_high <= 0:\n", str(len(invalid_high)))

Number of rows where range_low > range_high:
 121
Number of rows where range_low < 0 or range_high < 0:
 0


3. range is different for same manufacturer, omopid, and demographic

See below

4. unit is unclear ('not reported', 'index', 'units', 'index value', '(calc)', 'nan', None)
    - DONE Result: 372,650 rows

In [ ]:
# 4. unit is unclear
unclear = ['not reported', 'index', 'units', 'index value', '(calc)', 'nan', None]

uncleardf = filtered[filtered['unit'].isin(unclear)]
# append unclear indeces to to_drop
to_drop += list(uncleardf.index)

nanDF = filtered[filtered['unit'].isnull()]
# append unclear indeces to to_drop
to_drop += list(nanDF.index)

print("Number of rows with unclear units (''not reported', 'index', 'units', 'index value', '(calc)', 'nan', None):")
print(len(uncleardf) + len(nanDF))

Number of rows with unclear units (''not reported', 'index', 'units', 'index value', '(calc)', 'nan', None:
372650


5. age is negative
    - DONE Result: 90 rows

In [ ]:
# 5. age is negative
negAge = filtered[filtered['age'] < 0]
# Create list of indeces to drop from filtered data frame. Will append as we check cases
to_drop += list(negAge.index)

print("Number of rows with age value < 0:")
print(len(negAge))
negAge

Number of rows with age value < 0:
90


OMOPid manufacturer     unit  range_low  range_high gender  \
1295363   3018572      LAB3003   mmol/l       98.0       108.0   Male   
1295365   3013290      LAB3003    mm hg       35.0        45.0   Male   
1295396   3027315      LAB3003    mm hg       80.0       100.0   Male   
2269900   3033575       LAB293  x10e9/l        0.1         0.6   Male   
2269901   3000905       LAB293  x10e9/l        3.4        10.0   Male   
...           ...          ...      ...        ...         ...    ...   
14170110  3013502      LAB3239        %       95.0        99.0   Male   
14170111  3027315      LAB3239    mm hg       80.0       100.0   Male   
14170112  3013290      LAB3239    mm hg       27.0        41.0   Male   
14170113  3024882      LAB3239        %       20.0       100.0   Male   
14170114  3013290      LAB3239    mm hg       27.0        41.0   Male   

                       ethnicity                race  age  LOINCid  
1295363       Hispanic or Latino             Unknown  -99   2069-3  
1295365       Hispanic or Latino             Unknown  -99  11557-6  
1295396       Hispanic or Latino             Unknown  -99  11556-8  
2269900   Not Hispanic or Latino  White or Caucasian   -5    742-7  
2269901   Not Hispanic or Latino  White or Caucasian   -5   6690-2  
...                          ...                 ...  ...      ...  
14170110  Not Hispanic or Latino               Asian   -1  20564-1  
14170111  Not Hispanic or Latino               Asian   -1  11556-8  
14170112  Not Hispanic or Latino               Asian   -1  11557-6  
14170113  Not Hispanic or Latino               Asian   -1  19994-3  
14170114  Not Hispanic or Latino               Asian   -1  11557-6  

[90 rows x 10 columns]

In [ ]:

to_drop = list(set(to_drop))
print("Dropping " + str(len(to_drop)) + " rows from df")

Dropping 372730 rows from df


In [ ]:
exp = len(filtered) - len(to_drop)

In [ ]:
# drop unclear rows from DF
# toDrop = list(set(errRange.index.append(negAge.index).append(uncleardf.index).append(nanDF.index)))
unclear_range = filtered.loc[to_drop]

# save unclear ranges df to csv:
# unclear_range.to_csv('unclearOMOPRanges.csv')

# filter out all these unclear dfs from main df.
clear_ranges = filtered.drop(to_drop, axis=0,inplace=False)

In [ ]:
post_drop_length = len(clear_ranges)
print("Dropping unclear values...\nSuccess: ", post_drop_length == exp)

Dropping unclear values...
Success:  True


In [ ]:
subsample = clear_ranges.head(1000)

# Converting Units
## Steps:
### 1. Implement functions to get conversion map (unit_map) for each lab
      a) getMaxUnit(df, c)
      b) getUnits(df)
### 2. Create dict with conversion multipliers for each conversion necessary
### 3. Implement convertUnits(df)





Want to see if units within a same lab are in a different scale.
Brute force approach would be to make constant conversions across the board.

However, we shouldn't apply conversions to all the data at once because then we might end up with values such as 95,000,000 or 0.000000095 (i.e. cant generalize which unit to convert to across labs, since for one lab it might make more sense to have the ranges in micrograms, while in another lab it might make more sense to have the unit be in milligrams).


Need to check within each lab, which values to convert. This way we don't end up with very large or very small values, which would make us have to do another round of conversion when comparing individual results to these ranges.

The functions I implemented to accomplish this task of converting units are:

- getMaxUnit(df, c)
  - inputs a df and a cutoff value.
  - returns whether the max unit is frequent enough, the list of units present, and the max unit
- getUnits(df)
  - inputs a df
  - calls getMaxUnit(df, c)
  - returns a dictionary that maps each OMOPid to its max unit and the list of units present if a max unit is present that meats the cutoff value.
- convertUnits(df)
  - inputs a df
  - returns a df with the conversions applied


# Examples of randomness in units used for the same lab
Format is as follows:

OMOPid: [units used for ranges in this lab]



***Units that clearly don't match, unclear how to convert.***
- Approach: Look at how dominant each unit is, set threshold to delete certain labs that don't meet it.

3009744: ['g/dl', '%']

3008625: ['pg/ml', 'mcg/g creat']

3011337: ['ng/dl', 'ug/24 hr']

3010493: ['relative to h2o', 'rel.saline']

3012870: ['mg/24 hr', 'ratio']

3014521: ['mcg/g creat', 'ug/d']

3017575: ['u/ml', 'mg/dl', 'ng/dl', '%', 'u/l', 'mmol/l']

3018478: ['mcg/24 h', 'ug/l']

3020027: ['mg/24 h', 'mg in tv', 'mg/d']

3020416: ['x10e6/ul',
  '10*6/ul',
  'm/ul',
  'x10e6',
  'x10e6 cells/ul',
  'm/mm3',
  'm/cmm',
  'x10e12/l',
  'million/ul',
  '10e6/mm3',
  '10^6/\\xb5l']
  
3024929: ['thousand/ul',
  '10e3/mm3',
  'x10e3',
  'k/mm3',
  'x10e3 cells/ul',
  'x10e3/ul',
  'k/\\xb5l',
  't/mm3',
  'k/ul',
  '10*3/ul',
  '100k',
  'x10e9/l']
  
3031400: ['ug/l', 'mcg/spec']

3033575: ['10e3/mm3',
  'x10e3',
  'x10e3 cells/ul',
  'x10e3/ul',
  'k/ul',
  'cells/ul',
  '10*3/ul',
  'x10e9/l']
  
  
3037121: ['mg/24 hr', 'mg/dl']
  
  
3038191: ['umol/d', 'nm/mg cr']


3041450: ['nmol/min/ml', 'ng/ml']
  


  



***Examples of abnormal unit groups, diffcult to use regex on since differences are so unique:***

3000067: ['ng/l', 'pg/ml']

3000120: ['mg/24 h', 'mg in tv']

3000620: ['mg/dl adult', 'mg/dl']

3000905: ['thousand/ul',
  'x10e3/ul',
  'x10e3',
  '10e3/mm3',
  'x10e3 cells/ul',
  'x10e9/l',
  'k/mm3',
  'k/ul',
  '10*3/ul',
  'th/mm3']
  
3001218: ['mg', 'mg/d', 'mg/24 h', 'mg/24 hr']

3001405: ['/ul', 'cells/ul', 'x10e6/l']

3001754: ['ug/dl rbcs', 'ug/dl rbc']

3004239: ['g in tv', 'g/d', 'mg/d', 'mg/24 hr', 'mg/kg/d', 'g/24 h']

3004327: ['x10e3',
  'th/mm3',
  'k/mm3',
  'x10e3 cells/ul',
  'x10e3/ul',
  'k/ul',
  'cells/ul',
  '10*3/ul',
  'x10e9/l']
  
3008295: ['mosmol/kg', 'mmol/kg', 'mosm/kg']

3010493: ['relative to h2o', 'rel.saline']

3023851: ['mcg/24 h', 'mcg/24h', 'ug/24 hr']

3006923: ['u/l', 'iu/l']

3007220: ['u/l', 'iu/l'],

3009201: ['miu/l', 'uiu/ml']

3017526: ['ug/l', 'mcg/g creat']




***Examples of obvious changes***

 3002022: ['percent', '%']

 3002109: ['ratio units', 'ratio']

 3003308: ['% activity', '%']

 3008295: ['mosmol/kg', 'mmol/kg', 'mosm/kg']

 3008377: ['nmol/ml', 'umol/l']

 3016948: ['sec', 's']



## 1. Implement functions to get conversion map (unit_map) for each lab
    a) getMaxUnit(df, c)
    b) getUnits(df)

In [ ]:

# a) getMaxUnit(df, c)
# c is the percentage cutoff we want to use to determine whether the most frequent unit is really a significant majority of the data.
# for example, at the last if statement in this code, it checks whether the amount of times the most frequent unit appeared is greater than the total number of observations * c.
# So if c = 0.7, it would only count this as the most frequent unit if it has a frequency greater or equal to 70%.
# We later experiment with different values of c to see how many clear lab ranges we can determine.
def getMaxUnit(df, c):
    # get unit counts
    counts_obj = df['unit'].value_counts()
    counts = list(counts_obj)
    maxv = 0
    tot = 0
    index = 0
    # loop through counts and determine the frequency of the most frequent unit
    for i in range(len(counts)):
        count = counts[i]
        if count > maxv:
            maxv = count
            index = i
        tot += count
    # if it is frequent enough, return True, the list of existing units in this df, and the max unit
    if maxv >= (tot*c):
        return True, list(counts_obj.index), list(counts_obj.index)[index]
    else:
        return False, None, None

In [ ]:
from itertools import compress

# b) getUnits(df)
# Remember, entries for the same lab tend to have different units since they come from different manufacturers, and medical professionals are also just not consistent with it within the same manufacturer.
# the purpose of this function is to take a df, and for each lab (OMOPid) find the max unit and the list of observed units used in this lab. With this, we will be able to make the conversions to the max Unit for each lab,
# depending on the max unit for this lab, solving the issue that would arise (as described earlier) if we took a brute force approach to converting units (making the same conversion regardless of the most frequent unit).
def getUnits(df):
    # get unique units for each lab
    units_obj = df.groupby('OMOPid').apply(lambda x: list(set(x['unit'])))
    units = list(units_obj)

    # get list of labs
    omops = list(units_obj.index)

    # intialize conversion dict
    omopToUnit = {}
    # get max unit info
    for i in range(len(omops)):
        if len(units[i]) < 2:
            continue
        lab = df[df['OMOPid'] == omops[i]]
        # if a max unit exists (frequency greater than 0.55), add to conversion dict
        tf, curr_units, maxUnit = getMaxUnit(lab, 0.55)
        if tf:
            omopToUnit[omops[i]] = [maxUnit, curr_units]

    return omopToUnit

unit_map = getUnits(clear_ranges)

In [ ]:
# this is the output from getUnits(clear_ranges).
# it maps each lab to its most frequent unit (if it made the cutoff. those that did not make the cutoff are considered too ambiguous and disregarded), and to a list of present units
# this map will be used to make conversions
unit_map

# 2. Create dict with conversion multipliers for each conversion necessary

In [ ]:
# Due to the inconsistent nature of the data, it was too difficult to automate the process of getting the conversions between two units.
# There were too many typos, and mostly ambiguous abbreviations that made it difficult for anything but a human eye with lab test knowledge to discern what different abbreviations likely meant.
# thus the conversions were hard coded, all from the unit map above.
conversions = {'mg/24 hr': {'mg/d' : 1},
               'mg/24 h' : {'mg/d':1},
               'ug/24 hr' : {'mcg/24 h' : 1},
               'g/24 hr' : {'g/d' : 1},
               'g/24 h' : {'g/d' : 1},
               'mm/hr' : {'mm/h' : 1},
               'mmol/24 hr' : {'mmol/d': 1},
               'mmol/24 h' : {'mmol/d' : 1},

               'ug feu/ml' : {'mg/l feu': 1},

               'million/ul' : {'x10e12/l' : 1},
               'x10e6/ul' : {'x10e12/l' : 1},
               'x10e6 cells/ul' : {'x10e12/l' : 1},
               'x10e6/ul' : {'x10e12/l' : 1},
               '10e6/mm3' : {'x10e12/l' : 1},
               'm/ul' : {'x10e12/l' : 1},

               'mosm/kg' : {"mmol/kg" : 1},
               'mosmol/kg' : {"mmol/kg" : 1},
               'meq/l' : {'mmol/l' : 1},
               'eu/dl' : {'mg/dl' : 1},
               'mm/l' : {'mmol/l' : 1},
               'umol/l' : {'nmol/ml' : 1},
               "nmol/ml" : {'umol/l' : 1},



               'ng/ml/hr' : {'ng/ml/h' : 1},
               'units/l' : {'u/l' : 1},
               'sec' : {'s' : 1},

               'mg/g creat' : {'mg/g cr' : 1},
               'mg/mg creat' : {'mg/g creat' : 1000},
               'ug/mg creat' : {'mg/g creat' : 1},

               'pg/ml' : {'ng/ml': 0.001, 'ng/l' : 1},
               'ng/ml' : {'mcg/l' : 1, 'ug/l' : 1, 'ug/l rbc' :1 , 'ug/ml' : .001, 'mg/dl' : .0001, 'pg/ml' : 1000},
               'ug/ml' : {'mcg/ml' : 1, 'mg/ml' : .001, 'mg/l' : 1, 'mg/dl' : .1},
               'ug/dl' : {'mcg/dl' : 1},
               'mg/dl (calc)' : {'mg/dl' : 1},
               'mg/dl' : {'g/dl' : .001},
               'mg/dl adult' : {'mg/dl' : 1},
               'mcg/dl' : {'ug/dl' : 1},
               'g/dl' : {'g/dl (calc)' : 1},
               'mg/l' : {'ng/ml' : 1000},
              'ug/l' : {'ng/ml' : 1, 'mcg/l' : 1, 'mcg/dl' : 0.1},
               'u/l' : {'mcg/l' : 1},

               'mm[hg]' : {'mm hg' : 1},
               'mm b hg' : {'mm hg' : 1},


              '% (calc)' : {'%' : 1},
              'percent' : {'%' : 1},
              '% normal' : {"%" : 1},
              '%activity' : {"%" : 1},
              '% activity' : {"%" : 1},
              '%nhp' : {"%" : 1},
              '% inhibition' : {"%" : 1},
              '% nhp' : {"%" : 1},
              '% act' : {"%" : 1},
              '% of total' : {"%" : 1},
              '% lymphs' : {"%" : 1},


               'iu/l' : {'u/l' : 1},
               'uiu/ml' : {'miu/l' : 1},
               'miu/ml' : {'iu/l' : 1},


              'thousand/ul' : {'x10e9/l' : 1},
              'x10e3/ul' : {"x10e9/l" : 1},
              'x10e3 cells/ul' : {"x10e9/l" : 1},
              'k/ul' : {"x10e9/l" : 1, 'x10e3/ul' : 1},
              'k/mm3': {"x10e9/l" : 1, 'x10e3/ul' : 1},
              '10e3/mm3' : {'x10e9/l' : 1},
              '10*3/ul' : {'x10e9/l' : 1},
              'th/mm3' : {'x10e9/l' : 1},

               'cells/ul' : {'x10e6/l' : 1, 'x10e9/l' : 0.001, '/ul' : 1},
               'cells/mcl': {'cells/ul' : 1},

              '/ul' : {'x10e6/l' : 1}}






# 3. Implement convertUnits(df)

In [ ]:
from statistics import mode

# converts units in a df for a lab to the most frequent unit to ensure all ranges are of same unit
# returns a the df with the conversions applied.

def convertUnits(df):
    units = list(df['unit'])
    maxUnit = mode(units)

    # df where unit is already most common
    keep = df[df['unit'] == maxUnit]

    # df where unit can be converted to the most common
    mid = df[(df['unit'].isin(conversions)) & (df['unit'] != maxUnit)]

    multipliers = []

    for unit in list(mid['unit']):
      # check if conversion exists
        if maxUnit in conversions[unit]:
          # if so, append the value for this conversion
            multipliers.append(conversions[unit][maxUnit])
        # if there is no logical conversion (ex: 'mg/24 hr' to 'mg/dl',
        # make multiplier negative to then easily filter out of df
        else:
            multipliers.append(-1)
    # apply changes.
    mid['range_low'] *= multipliers
    mid['range_high'] *= multipliers
    mid['unit'] = maxUnit

    # drop rows where range is negative, which will be those in which the multiplier was set to -1 to mark impossible conversions
    converted = mid[mid["range_high"] >= 0]


    # return df with rows converted to the most frequent unit
    return keep.append(converted)

# Convert units functionality is now complete. We will apply it later.
# First, lets do more exploration and cleaning...

We see many of the units are percentages, so we can consolidate these as well.

In [ ]:
percent_df2 = clear_ranges[clear_ranges["unit"].str.contains("%")]
set(percent_df2['unit'])

{'%',
 '% (calc)',
 '% act',
 '% activity',
 '% inhibition',
 '% lymphs',
 '% nhp',
 '% normal',
 '% of total',
 '%activity',
 '%nhp',
 'vol %',
 'vol%',
 'vols%'}

In [ ]:
## change all units that contain "%" to just "%"
clear_ranges.loc[clear_ranges["unit"].str.contains("%"), "unit"] = "%"
clear_ranges.replace("percent", "%", inplace=True)

In [ ]:
# get a smaller subset to work on this algorithm
labSubset = clear_ranges[(clear_ranges['OMOPid'] == 3024561) & (clear_ranges['gender'] == 'Female') & (clear_ranges['age'] >= 30) & (clear_ranges['age'] < 40) & (clear_ranges['ethnicity'] == "Not Hispanic or Latino") & (clear_ranges['race'] == "White or Caucasian") & (clear_ranges['age'] == 34)]


Remember, the end-goal involves being able to compare patient results to these ranges to determine whether the result is normal. How could we distinguish between ranges 3.5-4.8 vs 3.4-4.7?

Here we see there is a need for a confidence interval. We will implement one to merge similar ranges for a lab that have matching demographics.

In [ ]:
# shows need for confidence interval (DONE)
print("differing ranges for same omop and demographics.")
labSubset.drop_duplicates(subset = ['range_low', 'range_high'])

differing ranges for same omop and demographics.


OMOPid manufacturer   unit  range_low  range_high  gender  \
10        3024561      LAB5389   g/dl        3.5         4.8  Female   
569264    3024561        LAB45   g/dl        3.4         4.7  Female   
2039524   3024561      LAB6170   g/dl        3.5         5.0  Female   
2195584   3024561      LAB3206  mg/dl     3600.0      4800.0  Female   
2230774   3024561      LAB6170   g/dl        3.8         4.8  Female   
5248626   3024561      LAB6170   g/dl        3.6         5.1  Female   
12461477  3024561      LAB3206  mg/dl     3503.0      5000.0  Female   

                       ethnicity                race  age LOINCid  
10        Not Hispanic or Latino  White or Caucasian   34  1751-7  
569264    Not Hispanic or Latino  White or Caucasian   34  1751-7  
2039524   Not Hispanic or Latino  White or Caucasian   34  1751-7  
2195584   Not Hispanic or Latino  White or Caucasian   34  1751-7  
2230774   Not Hispanic or Latino  White or Caucasian   34  1751-7  
5248626   Not Hispanic or Latino  White or Caucasian   34  1751-7  
12461477  Not Hispanic or Latino  White or Caucasian   34  1751-7

Here we have another example, but even with **the same manufacturer**. This emphasizes how inconsistent the EHR system is, but as far as it concerns our study, it strongly emphasizes the need for a confidence interval to merge these similar ranges.

In [ ]:
# Another example, but even with same manufacturer. Emphasizes need for confidence interval (DONE)
print('Example of discordant ranges for rows with same lab (OMOPid), manufacturer, and demographics')
labSubset[labSubset['manufacturer'] == 'LAB45'].drop_duplicates(subset = ['range_low', 'range_high'])

Example of discordant ranges for rows with same lab (OMOPid), manufacturer, and demographics


OMOPid manufacturer  unit  range_low  range_high  gender  \
1101     3024561        LAB45  g/dl        3.5         4.8  Female   
569264   3024561        LAB45  g/dl        3.4         4.7  Female   
3388387  3024561        LAB45  g/dl        3.5         5.0  Female   
8317838  3024561        LAB45  g/dl        3.6         5.1  Female   

                      ethnicity                race  age LOINCid  
1101     Not Hispanic or Latino  White or Caucasian   34  1751-7  
569264   Not Hispanic or Latino  White or Caucasian   34  1751-7  
3388387  Not Hispanic or Latino  White or Caucasian   34  1751-7  
8317838  Not Hispanic or Latino  White or Caucasian   34  1751-7

# Ok, so in order to do further consolidation. We need to:

  a) **merge similar ranges** using a confidence interval


    merge_similar_ranges(df, ci)

  b) **merge demographics**. That is, set demographic values to 'any' in places where, for example, the range is the same regardless of gender. In this case the gender column should say 'any'.


    merge_demographics(df)

In [ ]:
clear_ranges

OMOPid manufacturer            unit  range_low  range_high  gender  \
1         3013502      LAB5808               %      95.00       99.00  Female   
2         3013502      LAB5809               %      95.00       99.00  Female   
3         3027315      LAB5808           mm hg      80.00      100.00  Female   
4         3027315      LAB5809           mm hg      80.00      100.00  Female   
5         3006576      LAB5808          mmol/l      22.00       27.00  Female   
...           ...          ...             ...        ...         ...     ...   
14244781  3016723      LAB3001           mg/dl       0.61        1.24    Male   
14244782  3000905       LAB294         x10e9/l       3.40       10.00  Female   
14244783  3046553       LAB293         x10e9/l       0.00        0.20  Female   
14244784  3004327       LAB293         x10e9/l       0.90        2.90    Male   
14244785  3023996       LC3743  mmol/mol creat       1.00       14.00    Male   

                       ethnicity                race  age  LOINCid  
1         Not Hispanic or Latino               Asian    0  20564-1  
2         Not Hispanic or Latino               Asian    0  20564-1  
3         Not Hispanic or Latino               Asian    0  11556-8  
4         Not Hispanic or Latino               Asian    0  11556-8  
5         Not Hispanic or Latino               Asian    0   1959-6  
...                          ...                 ...  ...      ...  
14244781      Hispanic or Latino  White or Caucasian   53   2160-0  
14244782                 Unknown             Unknown   71   6690-2  
14244783      Hispanic or Latino  White or Caucasian    2  43743-4  
14244784  Not Hispanic or Latino  White or Caucasian   36    731-0  
14244785      Hispanic or Latino             Unknown   13  25132-2  

[13872056 rows x 10 columns]

In [ ]:
df = clear_ranges

In [ ]:
#define our own age ranges
ages = [[0,10], [10,20], [20,30], [30,40], [40,50], [50,60], [60,70], [70,80], [80,90], [90,100]]

# create list for each possible demographic
races = list(set(df['race']))
races.remove('Unknown')

ethnicities = list(set(df['ethnicity']))
ethnicities.remove('Unknown')

genders = list(set(df['gender']))

In [ ]:
ethnicities

['Hispanic or Latino', 'Not Hispanic or Latino']

# A) Merge similar ranges

In [ ]:
def merge_similar_ranges(df, ci = float):
  '''Merges similar ranges for a given lab (df), given a confidence interval (ci), in place'''
    vis = []
    range_CI = {}
    # remember to consider that one range might be mapped to another, which is mapped to a different one. Might have to merge all in this case.
    # To overcome this obstacle, get the mappings between ranges that are similar enough to merge into the same range
    for i, row in df.iterrows():
        index = row
        low = row['range_low']
        high = row['range_high']

        curr_interval = str(low)+'+'+str(high)

        # if we saw this range already, continue. If we haven't, go through and see if it falls in the CI of any visited range
        if [low, high] in vis:
            new_range = range_CI[curr_interval]
            df.at[i,'range_low'] = new_range[0]
            df.at[i,'range_high'] = new_range[1]
            continue

        # initially, map this range to itself in case it doesn't end up in the CI of another range.
        range_CI[curr_interval] = [low, high]

        # get the lower and upper bound intervals for the low and high values of the range
        low_int = [low-low*ci, low+low*ci]
        high_int = [high-high*ci, high+high*ci]

        # iterate through unique ranges we already visited
        for interval in vis:
            # if this range falls in the CI of one of the ranges we already saw
            if (interval[0] >= low_int[0]) and (interval[0] <= low_int[1]) and (interval[1] >= high_int[0]) and (interval[1] <= high_int[1]):

                #map the current range to this one
                range_CI[curr_interval] = [interval[0],interval[1]]

                # change the value of this range to match the other
                df.at[i,'range_low'] = interval[0]
                df.at[i,'range_high'] = interval[1]
            break

        if [low,high] not in vis:
            vis.append((low,high))

    return


# Now lets test merge_similar_ranges() on a sample lab and check if it works.

In [ ]:
# original lab sample, good for testing code below:
# to test with pdfs:
sample_lab = df[df['LOINCid'] == '789-8']
x = sample_lab

In [ ]:
x

OMOPid manufacturer      unit  range_low  range_high  gender  \
90        3020416       LAB294  x10e12/l        4.0         5.2  Female   
91        3020416       LAB294  x10e12/l        4.0         5.2  Female   
92        3020416       LAB294  x10e12/l        4.0         5.2  Female   
93        3020416       LAB294  x10e12/l        4.0         5.2  Female   
94        3020416       LAB294  x10e12/l        4.0         5.2  Female   
...           ...          ...       ...        ...         ...     ...   
14244533  3020416       LAB294  x10e12/l        4.0         5.2  Female   
14244550  3020416       LAB294  x10e12/l        4.4         5.9    Male   
14244669  3020416       LAB293  x10e12/l        4.4         5.9    Male   
14244731  3020416       LAB293  x10e12/l        4.2         5.6    Male   
14244778  3020416       LAB293  x10e12/l        4.0         5.2  Female   

                       ethnicity                race  age LOINCid  
90        Not Hispanic or Latino  White or Caucasian   35   789-8  
91        Not Hispanic or Latino  White or Caucasian   35   789-8  
92        Not Hispanic or Latino  White or Caucasian   35   789-8  
93        Not Hispanic or Latino  White or Caucasian   35   789-8  
94        Not Hispanic or Latino  White or Caucasian   35   789-8  
...                          ...                 ...  ...     ...  
14244533  Not Hispanic or Latino             Unknown   38   789-8  
14244550  Not Hispanic or Latino  White or Caucasian   33   789-8  
14244669  Not Hispanic or Latino               Asian   22   789-8  
14244731                 Unknown  White or Caucasian   16   789-8  
14244778                 Unknown  White or Caucasian   70   789-8  

[787184 rows x 10 columns]

In [ ]:
sample1 = x[x['range_low'] == 4].sample(50)
sample2 = x[x['range_low'] == 4.14].sample(50)
sample3 = x[x['range_low'] == 3.8]

In [ ]:
# get a sample of data for each range from a lab
z = sample1.append(sample2).append(sample3)

<ipython-input-76-de6316c891f3>:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  z = sample1.append(sample2).append(sample3)
<ipython-input-76-de6316c891f3>:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  z = sample1.append(sample2).append(sample3)


In [ ]:
z

OMOPid manufacturer        unit  range_low  range_high  gender  \
12425516  3020416       LAB293    x10e12/l        4.0         5.2  Female   
9010898   3020416       LAB294    x10e12/l        4.0         5.2  Female   
11036010  3020416       LAB294    x10e12/l        4.0         5.2  Female   
6832396   3020416       LAB294    x10e12/l        4.0         5.2  Female   
2069836   3020416       LAB293    x10e12/l        4.0         5.2  Female   
...           ...          ...         ...        ...         ...     ...   
14238589  3020416       LAB293  million/ul        3.8         5.1  Female   
14240514  3020416       LAB293  million/ul        3.8         5.1  Female   
14243171  3020416       LAB293  million/ul        3.8         5.1  Female   
14244120  3020416       LAB293  million/ul        3.8         5.1  Female   
14244444  3020416       LAB294  million/ul        3.8         5.1  Female   

                       ethnicity                                       race  \
12425516  Not Hispanic or Latino                  Black or African American   
9010898                  Unknown                                    Unknown   
11036010  Not Hispanic or Latino                         White or Caucasian   
6832396   Not Hispanic or Latino  Native Hawaiian or Other Pacific Islander   
2069836   Not Hispanic or Latino                         White or Caucasian   
...                          ...                                        ...   
14238589  Not Hispanic or Latino                         White or Caucasian   
14240514  Not Hispanic or Latino                         White or Caucasian   
14243171      Hispanic or Latino                                    Unknown   
14244120  Not Hispanic or Latino                         White or Caucasian   
14244444  Not Hispanic or Latino                         White or Caucasian   

          age LOINCid  
12425516   33   789-8  
9010898    44   789-8  
11036010   42   789-8  
6832396    55   789-8  
2069836    57   789-8  
...       ...     ...  
14238589   26   789-8  
14240514   54   789-8  
14243171   38   789-8  
14244120   59   789-8  
14244444   19   789-8  

[13373 rows x 10 columns]

In [ ]:
merge_similar_ranges(z, ci = 0.1)

### compare z before and after merge_similar_ranges()
#### note that the ci = 0.1, so when taking 4.0 as the low range, it looks for anything between 3.6 and 4.4, and merges the range to 4.0.
### Success!

In [ ]:
z

OMOPid manufacturer        unit  range_low  range_high  gender  \
12425516  3020416       LAB293    x10e12/l        4.0         5.2  Female   
9010898   3020416       LAB294    x10e12/l        4.0         5.2  Female   
11036010  3020416       LAB294    x10e12/l        4.0         5.2  Female   
6832396   3020416       LAB294    x10e12/l        4.0         5.2  Female   
2069836   3020416       LAB293    x10e12/l        4.0         5.2  Female   
...           ...          ...         ...        ...         ...     ...   
14238589  3020416       LAB293  million/ul        4.0         5.2  Female   
14240514  3020416       LAB293  million/ul        4.0         5.2  Female   
14243171  3020416       LAB293  million/ul        4.0         5.2  Female   
14244120  3020416       LAB293  million/ul        4.0         5.2  Female   
14244444  3020416       LAB294  million/ul        4.0         5.2  Female   

                       ethnicity                                       race  \
12425516  Not Hispanic or Latino                  Black or African American   
9010898                  Unknown                                    Unknown   
11036010  Not Hispanic or Latino                         White or Caucasian   
6832396   Not Hispanic or Latino  Native Hawaiian or Other Pacific Islander   
2069836   Not Hispanic or Latino                         White or Caucasian   
...                          ...                                        ...   
14238589  Not Hispanic or Latino                         White or Caucasian   
14240514  Not Hispanic or Latino                         White or Caucasian   
14243171      Hispanic or Latino                                    Unknown   
14244120  Not Hispanic or Latino                         White or Caucasian   
14244444  Not Hispanic or Latino                         White or Caucasian   

          age LOINCid  
12425516   33   789-8  
9010898    44   789-8  
11036010   42   789-8  
6832396    55   789-8  
2069836    57   789-8  
...       ...     ...  
14238589   26   789-8  
14240514   54   789-8  
14243171   38   789-8  
14244120   59   789-8  
14244444   19   789-8  

[13373 rows x 10 columns]

# B) Merge Demographics

In [ ]:
# Input a list of ages. Returns True if 50% of age bins are present (bin_size=10 would give 0-10, 10-20, ..., 90-100),
# indicating that we can replace the age column with 'any'
def all_ages(ages, bin_size = int):
    # loop through all age bins (0-10, 10-20, ..., 90-100)
    vals = []
    for i in range(0,100,bin_size):
        lower = i
        upper = i+bin_size
        # append True if there's an age in ages list that falls in the current bin
        vals.append(any([True for age in ages if age >= lower and age < upper]))

    if (sum(vals)/len(vals)) >= 0.50:
        return True

    return False


In [ ]:
ethnicities

['Hispanic or Latino', 'Not Hispanic or Latino']

In [ ]:
races

['White or Caucasian',
 'American Indian or Alaska Native',
 'Native Hawaiian or Other Pacific Islander',
 'Asian',
 'Black or African American']

In [ ]:
# Inputs a df for a lab with ranges already merged where necessary. Returns a df with 'any'
# in the demographic columns if the range does not depend on this demographic (determined on whether all possible
# values for a demographic are present for a given range). not in place.
def merge_demographics(df):

    # initialize df to append to during iteration
    output_df = pd.DataFrame(columns=list(df.columns))

    # break into new dfs per range, search to see if value can be replaced by 'any' for each range
    ranges = df[['range_low', 'range_high']].drop_duplicates()


    # iterate through unique ranges for this lab
    for i, lab_range in ranges.iterrows():
        low = lab_range['range_low']
        high = lab_range['range_high']

        # get all rows that have this specific range for this lab
        curr_df = df[(df['range_low'] == low) & (df['range_high'] == high)]


        # now check if the df for this range contains all values for each demographic and if so, change to 'any'
        # then append all rows to the new table

        # check if all values for gender are present
        curr_gender_vals = list(set(curr_df['gender']))
        if len(genders) == len(curr_gender_vals):
            curr_df['gender'] = 'any'

        # check if all values for ethnicity are present
        curr_ethnicity_vals = list(set(curr_df['ethnicity']))
        # value could be "Unknown". check if all defined ethnicities are in the current range
        if all([x in curr_ethnicity_vals for x in ethnicities]):
            curr_df['ethnicity'] = 'any'

        # check if all values for race are present
        curr_race_vals = list(set(curr_df['race']))
        # value could be "Unknown". check if all defined races are in the current range
        if all([x in curr_race_vals for x in races]):
            curr_df['race'] = 'any'

        # get age in intervals:
        curr_df['age_range'] = [ages[age//10] for age in curr_df['age']]

        # get first digit of age range to drop duplicates on later (can't drop duplicates on age_range because it's a list)
        curr_df['age_index'] = [age_range[0] for age_range in curr_df['age_range']]

        # check if all values for age (range) are present
        if all_ages(list(set(curr_df['age'])), bin_size = 30):
            curr_df['age'] = 'any'

        # drop duplicates
        curr_df.drop_duplicates(subset = (['gender', 'ethnicity', 'race', 'age_index']), inplace = True)
        # no longer need age_index col, drop it
        curr_df.drop('age_index', inplace=True, axis = 1)


        #concatenate result to output
        output_df = output_df.append(curr_df)

    return output_df


# Lets try merge_demographics() on our sample dataset...

In [ ]:
# merge_demographics() on sample data set from before: z
output = merge_demographics(z)

In [ ]:
# print the output, which is basically the concatenated dfs for each range with demographic set to 'any' where it meets the condition
output

OMOPid manufacturer        unit  range_low  range_high  gender  \
12425516  3020416       LAB293    x10e12/l       4.00         5.2     any   
9010898   3020416       LAB294    x10e12/l       4.00         5.2     any   
6832396   3020416       LAB294    x10e12/l       4.00         5.2     any   
4088755   3020416       LAB294    x10e12/l       4.00         5.2     any   
12914164  3020416       LAB294    x10e12/l       4.00         5.2     any   
1349612   3020416       LAB294    x10e12/l       4.00         5.2     any   
4196126   3020416       LAB293    x10e12/l       4.00         5.2     any   
13331515  3020416       LAB293    x10e12/l       4.00         5.2     any   
5739968   3020416       LAB294    x10e12/l       4.00         5.2     any   
2384990   3020416       LAB293  million/ul       4.00         5.2     any   
12709954  3020416       LAB294    x10e12/l       4.00         6.6  Female   
5148665   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
9476129   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
7996663   3020416       LAB294    x10e6/ul       4.14         5.8    Male   
8342842   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
14051799  3020416       LAB294    x10e6/ul       4.14         5.8    Male   
8120200   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
3493123   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
7684824   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
8328433   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
10851506  3020416       LAB294    x10e6/ul       4.14         5.8    Male   
12949315  3020416       LAB293    x10e6/ul       4.14         5.8    Male   
3334528   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
12675998  3020416       LAB293    x10e6/ul       4.14         5.8    Male   
9589599   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
8262285   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
12240673  3020416       LAB293    x10e6/ul       4.14         5.8    Male   
344000    3020416       LAB293    x10e6/ul       4.14         5.8    Male   
13889781  3020416       LAB293    x10e6/ul       4.14         5.8    Male   
2504522   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
9655921   3020416       LAB293    x10e6/ul       4.14         5.8    Male   
1894956   3020416       LAB293     10*6/ul       3.80         5.9    Male   
8884081   3020416       LAB293     10*6/ul       3.80         5.9    Male   
9888944   3020416     LAB11007     10*6/ul       3.80         5.9    Male   
11325452  3020416       LAB293     10*6/ul       3.80         5.9    Male   

                       ethnicity                       race  age LOINCid  \
12425516                     any                        any  any   789-8   
9010898                      any                        any  any   789-8   
6832396                      any                        any  any   789-8   
4088755                      any                        any  any   789-8   
12914164                     any                        any  any   789-8   
1349612                      any                        any  any   789-8   
4196126                      any                        any  any   789-8   
13331515                     any                        any  any   789-8   
5739968                      any                        any  any   789-8   
2384990                      any                        any  any   789-8   
12709954                 Unknown                    Unknown    0   789-8   
5148665                      any         White or Caucasian  any   789-8   
9476129                      any                      Asian  any   789-8   
7996663                      any  Black or African American  any   789-8   
8342842                      any         White or Caucasian  any   789-8   
14051799              

# Great. With these methods the lab ranges can be used as reference for individual patient results. Now lets run it on all the data.

We will use clear_ranges as input, the following code will
- loop through each lab (OMOPid)
  - Check if there is only one unit
    - if so:
      - merges similar ranges
      - merges demographics
      - appends to the output
      - continues
    - if not:
      - checks if the units in this lab are convertible
      - if not, continues
  - if its made it this far, it means we have multiple units and they are convertible. proceeds to:
    - convert units
    - merge similar ranges
    - merge demographics
    - append to the output

In [ ]:
clear_ranges

OMOPid manufacturer            unit  range_low  range_high  gender  \
1         3013502      LAB5808               %      95.00       99.00  Female   
2         3013502      LAB5809               %      95.00       99.00  Female   
3         3027315      LAB5808           mm hg      80.00      100.00  Female   
4         3027315      LAB5809           mm hg      80.00      100.00  Female   
5         3006576      LAB5808          mmol/l      22.00       27.00  Female   
...           ...          ...             ...        ...         ...     ...   
14244781  3016723      LAB3001           mg/dl       0.61        1.24    Male   
14244782  3000905       LAB294         x10e9/l       3.40       10.00  Female   
14244783  3046553       LAB293         x10e9/l       0.00        0.20  Female   
14244784  3004327       LAB293         x10e9/l       0.90        2.90    Male   
14244785  3023996       LC3743  mmol/mol creat       1.00       14.00    Male   

                       ethnicity                race  age  LOINCid  
1         Not Hispanic or Latino               Asian    0  20564-1  
2         Not Hispanic or Latino               Asian    0  20564-1  
3         Not Hispanic or Latino               Asian    0  11556-8  
4         Not Hispanic or Latino               Asian    0  11556-8  
5         Not Hispanic or Latino               Asian    0   1959-6  
...                          ...                 ...  ...      ...  
14244781      Hispanic or Latino  White or Caucasian   53   2160-0  
14244782                 Unknown             Unknown   71   6690-2  
14244783      Hispanic or Latino  White or Caucasian    2  43743-4  
14244784  Not Hispanic or Latino  White or Caucasian   36    731-0  
14244785      Hispanic or Latino             Unknown   13  25132-2  

[13872056 rows x 10 columns]

In [ ]:
def generateRanges(clear_ranges):
  output_df = pd.DataFrame(columns = clear_ranges.columns)

  for omopid in set(clear_ranges['OMOPid']):
      curr_df = clear_ranges[clear_ranges['OMOPid'] == omopid]

      # check if multiple units
      if (len(set(curr_df['unit'])) < 2):
          # merge similar ranges (+/- 10%)
          merge_similar_ranges(curr_df, 0.25)
          # change demographic to 'any' where applicable
          cleaned = merge_demographics(curr_df)
          # append to output
          output_df = output_df.append(curr_df)
          continue

      # check if there is a dominant unit
      if omopid not in unit_map:
          # don't append if no dominant unit
          continue

      # convert if necessary
      converted = convertUnits(curr_df)
      # merge similar ranges (+/- 10%)
      merge_similar_ranges(converted, 0.25)
      # change demographic to 'any' where applicable
      cleaned = merge_demographics(converted)
      # append to output
      output_df = output_df.append(cleaned)


In [ ]:
output_df = generateRanges(clear_ranges)

In [ ]:
rows = len(output_df)
output_df.to_csv('cleaned_ehr_' + str(rows) + '_rows.csv')

# Outputs for different parameters:
5,840,912 rows for
- merge_similar_ranges(..., ci = 0.10)
- all_ages(..., bin_size = 10)
- getMaxUnit(..., 0.75)

5,840,364 rows for
- merge_similar_ranges(..., ci = 0.25)
- all_ages(..., bin_size = 30)
- all_ages: threshold = 0.50
- getMaxUnit(..., 0.55)

In [2]:
import pandas as pd
cleaned_ehr = pd.read_csv(filepath + 'cleaned_ehr.csv')
cleaned_ehr.set_index("Unnamed: 0", inplace=True)
cleaned_ehr.index.name = "Index"
cleaned_ehr

<ipython-input-2-999ca6afe5c2>:2: DtypeWarning: Columns (9,11) have mixed types. Specify dtype option on import or set low_memory=False.
  cleaned_ehr = pd.read_csv(filepath + 'cleaned_ehr.csv')


OMOPid manufacturer    unit  range_low  range_high  gender  \
Index                                                                   
1107112   3006467      LAB3381  umol/l      31.26       84.11  Female   
5202898   3006467       LC4764  umol/l      35.60      101.60  Female   
6504574   3006467      LAB3381  umol/l      32.10       71.90  Female   
8245085   3006467      LAB3381  umol/l      31.26       84.11    Male   
13622773  3006467      LAB3381  umol/l      31.26       84.11    Male   
...           ...          ...     ...        ...         ...     ...   
14239825  3016699      LAB3020   mg/dl      70.00      139.00  Female   
14242860  3016699      LAB3020   mg/dl      70.00      139.00  Female   
14243816  3016699      LAB3020   mg/dl      70.00      139.00  Female   
14243911  3016699      LAB3020   mg/dl      70.00      139.00  Female   
14244323  3016699      LAB3020   mg/dl      70.00      139.00  Female   

                       ethnicity                                       race  \
Index                                                                         
1107112   Not Hispanic or Latino                         White or Caucasian   
5202898   Not Hispanic or Latino                  Black or African American   
6504574   Not Hispanic or Latino                         White or Caucasian   
8245085   Not Hispanic or Latino                         White or Caucasian   
13622773  Not Hispanic or Latino                                      Asian   
...                          ...                                        ...   
14239825  Not Hispanic or Latino                                    Unknown   
14242860  Not Hispanic or Latino                         White or Caucasian   
14243816  Not Hispanic or Latino                                      Asian   
14243911  Not Hispanic or Latino  Native Hawaiian or Other Pacific Islander   
14244323  Not Hispanic or Latino                         White or Caucasian   

         age  LOINCid age_range  
Index                            
1107112   19  30195-2       NaN  
5202898   38  30195-2       NaN  
6504574   14  30195-2       NaN  
8245085   58  30195-2       NaN  
13622773  68  30195-2       NaN  
...       ..      ...       ...  
14239825  41   1504-0       NaN  
14242860  20   1504-0       NaN  
14243816  33   1504-0       NaN  
14243911  44   1504-0       NaN  
14244323  28   1504-0       NaN  

[5840912 rows x 11 columns]

In [3]:
len(set(cleaned_ehr['LOINCid']))

868